Add transparent protocol detection to proxy

seanmonstar opened the pull request:
The proxy will now try to detect what protocol new connections are
using, and route them accordingly. Specifically:

- HTTP/2 stays the same.
- HTTP/1 is now accepted, and will try to send an HTTP/1 request
  to the target.
- If neither HTTP/1 nor 2, assume a TCP stream and simply forward
  between the source and destination.

Closes #130 
Closes #131 

----

For HTTP/1 requests, it's possible there isn't an authority attached (1.0 doesn't know about `Host`, and even 1.1 is [allowed](https://tools.ietf.org/html/rfc7230#section-5.4) to send `Host: \r\n` if it doesn't know it). In these cases, the socket's `SO_ORIGINAL_DST` will be used to adjust the request-target.

Some things that may need to be done before merging:

- [x] (https://github.com/runconduit/conduit/issues/181) Ensure the controller destination service can echo back an address when we use the `SO_ORIGINAL_DST`.
- [x] Decide if transparency should start with config flags to disable.
- [x] (https://github.com/runconduit/conduit/issues/187) Decide if HTTP/1 telemetry should be disabled until the controller can better handle the large number of unique paths.

briansmith commented:
Sean, I see there are some commented-out lines of code in the PR. Did you upload the right version of the PR? If so, then what's the purpose of including the commented-out lines?

olix0r commented on proxy/src/transparency.rs:
```diff
@@ -0,0 +1,688 @@
+        Box::new(self.connect.connect()
+            .map_err(|_e| io::Error::new(io::ErrorKind::Other, "blah")))
+    }
+}
```
this file feels large enough that it would benefit from being split into a directory with multiple files?

seanmonstar commented:
@briansmith some of them are commenting out an executor, since it's something we wanted to be able to add some context to logs, but cannot be used in this PR as hyper requires that the executor be a `Handle`. I can just kill those lines, they don't provide much other than a slight reminder to add them back once I modify hyper to allow any executor and not just a `Handle`...

seanmonstar commented on proxy/src/transparency.rs:
```diff
@@ -0,0 +1,688 @@
+    fn call(&self, req: Self::Request) -> Self::Future {
+        let mut req: http::Request<hyper::Body> = req.into();
+        req.extensions_mut().insert(self.srv_ctx.clone());
+
```
Some HTTP/1-specific proxy stuff happens in here, perhaps it'd be better to lift this higher up so as to not seem like simply hyper glue...

hawkw commented on proxy/src/bind.rs:
```diff
@@ -151,30 +157,40 @@ where
+
 
-impl<B> control::discovery::Bind for Bind<Arc<ctx::Proxy>, B>
+impl<C, B> Bind<C, B> {
```
Very minor, but: seems to me this should still have the `// ===== impl Bind =====` comment over it?

hawkw commented on proxy/src/inbound.rs:
```diff
@@ -95,14 +83,15 @@ where
-    fn bind_service(&mut self, addr: &SocketAddr) -> Result<Self::Service, Self::RouteError> {
-        debug!("building inbound client to {}", addr);
+    fn bind_service(&mut self, key: &Self::Key) -> Result<Self::Service, Self::RouteError> {
+        let &(ref addr, proto) = key;
```
TIOLI, but why not:
```rust
fn bind_service(&mut self, &(ref addr, proto): &Self::Key) -> Result<Self::Service, Self::RouteError> {
    ...
```

seanmonstar replied:
I personally find destructuring in method signatures to be distracting in most cases.

hawkw replied:
That's fair; I find not introducing intermediate identifiers slightly clearer, but it's a matter of taste!

hawkw commented on proxy/src/outbound.rs:
```diff
@@ -54,19 +55,25 @@ where
-        )
+                self.default_zone.as_ref().map(|s| s.as_ref()));
+
+            let proto = match req.version() {
```
TIOLI: replace with `impl From<http::Version> for Protocol`?

seanmonstar replied:
Yea, could be done, but as this is only ever done in one place currently, having the match inline feels more descriptive to me.

hawkw commented on proxy/src/outbound.rs:
```diff
@@ -80,11 +87,15 @@ where
+        key: &Self::Key,
     ) -> Result<Self::Service, Self::RouteError> {
-        debug!("building outbound client to {:?}", authority);
+        let &(ref authority, protocol) = key;
```
Once again, take it or leave it, but consider destructuring in the method params?

seanmonstar commented:
Since you may have looked last, I've made these updates:

- Broke up `src/transparency.rs` into `src/transparency/*.rs` files.
- Added a test for `CONNECT` requests, ensuring that we send back a 502 instead of some other wonky behavior.
- Changed `Protocol::detect()` to return `Option<Protocol>`, as @olix0r mentioned, we aren't actually detecting TCP, but rather failing to detect a set of known protocols.

While there are a lot of changes in the `tests/support/` directory, you can ignore those and just look at `tests/transparency.rs` to see some fairly succinct test cases for protocol transparency.

olix0r commented on proxy/src/ctx/transport.rs:
```diff
@@ -56,6 +56,29 @@ impl Server {
+            Some(orig_dst) => {
+                // If the original destination is actually the listening socket,
+                // we don't want to create a loop.
+                if same_addr(&orig_dst, &self.local) {
```
I'm pretty sure this is right but want to confirm: `local` is the _bound_ local addr and not the "listen addr".  i.e. the listen addr might be 0.0.0.0:1234 but the bound addr is going to be i.e. 127.0.0.1:1234

We'll only know the bound addr once we have a socket on the serverside, whereas the local addr is known when we call _listen_ on the server.

olix0r replied:
yes this is right:
```rust
        let local_addr = connection.local_addr().unwrap_or(self.listen_addr);
```

seanmonstar replied:
I moved this code from `Outbound`, but will need to re-read to make sure.

olix0r commented on proxy/src/transparency/server.rs:
```diff
@@ -0,0 +1,158 @@
+        }
+    }
+
+    pub fn serve(&self, connection: Connection, remote_addr: SocketAddr) {
```
This would benefit from a comment outlining the general story of what the server does.

olix0r commented on proxy/src/transparency/server.rs:
```diff
@@ -0,0 +1,158 @@
+                    Some(Protocol::Http1) => {
+                        trace!("transparency sniffed HTTP/1");
+
+                        let srv_ctx = ServerCtx::new(
```
tioli:

could organize this as

```rust
Some(proto) => {
    trace!("transparency sniffed {}", proto);

    let srv_ctx = ...;

    let io = ...;

    match proto {
        Http1 => Box::new(...),
        Http2 => Box::new(...),
    }
}

olix0r commented:
Before merging, you'll need to run `bin/update-proxy-deps-shas` so that docker builds pick up your changes to Cargo.lock
